projects/03-captions-api/requirements.md

```diff
@@ -0,0 +1,205 @@
+
+```json
+{
+  "captions": [
+    {
+      "id": 123,
+      "url": "",
+      "text": "",
+      "caption_url": "" 
+    },
+    {
+      "id": 124,
+      "url": "",
+      "text": "",
+      "caption_url: ""
+    }
+  ]
+}
+```
+
+where: 
+- `id` is the generated unique ID
+- `url` is the original URL of the image submitted by the user
+- `text` is the text that should be added on the image
+- `caption_url` is the URL of the generated image.
+
+### GET /captions/:id
+Will return the attributes for the caption with id `:id`
+```json 
+{
+  "caption": {
+    "id": 123,
+    "url": "http://image.url/meme.jpg",
+    "text": "caption text",
+    "caption_url": "http://example.com/images/meme.jpb" 
+  },
+}
+```
+
+### DELETE /captions/:id
+Will delete the caption with that ID and its associated files. 
+The response will be `:200` if the delete was successful. 
+
+## Instagram Captions
+
+### POST /captions/instagrams
+
+API client will do a request: 
+
+```
+POST /captions/instagram
+```
+with the following body: 
+
+```json
+{
+  "image": {
+    "type": "image",
+    "url": "http://image.url",
+    "text": "caption text",
+    "filter": "blackwhite",
+  }
+}
+```
+The response will the same as for the `POST /captions`.
+
+See details about images sizes [here](https://help.instagram.com/1631821640426723), you should crop the image accordingly with the image sizes. 
+
+##### `type` 
+
+`type` attribute can be one of the following strings: 
+- `image`
+- `color`
+- `gradient`
+
+When type is `image` this means that an image will be provided as URL with `url` attribute and the API should use that as background image.
+
+Example of request: 
+```json
+{
+  "image": {
+    "type": "image",
+    "url": "http://image.url",
+    "text": "caption text"
+  }
+}
+```
+
+When type is `color` this means that there will not be provided an image, but a `color` attribute in HEX format and that will be used to generate a background with that color: 
+
+```json
+{
+  "image": {
+    "type": "color",
+    "color": "#003166",
+    "text": "caption text"
+  }
+}
+```
+
+When type is `gradient` this means that there will be two attributes `start_color` and `end_color` which will provide two colors in HEX format and those will be used to generate a background
+
+```json
+{
+  "image": {
+    "type": "gradient",
+    "start_color": "#000000",
+    "end_color": "#003166",
+    "text": "caption text",
+  }
+}
+```
+
+will generate a gradient background from white to blue with the text in the middle. 
+
+##### `filters` 
+
+Filters will be a list of filters that can be applied to an image before adding the text.
+
+The supported filters will be:
+- `blackwhite` - will make the image black and white, this should be allowed only when `type` is `image`
+- `light_blur` - will add a small blur the image, this should be allowed only with `type` = `image`
+- `hard_blur` - will add a deep blur to the image, this should be allowed only with `type` = `image`
```

Comment:
These filters will need to be applied to the image from the caption_url ?

Reply:
THe filters will be applied before adding the text for the generated image.

So the flow should be: 
1. Apply filter
2. Position the text
3. Save the new image